src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -210,12 +230,12 @@ private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int leng
                     return -1;
                 }
 
-                lock (value2metrics)
+                lock (this.tagsToMetricPointIndexDictionary)
                 {
                     // check again after acquiring lock.
-                    if (!value2metrics.TryGetValue(tagValues, out aggregatorIndex))
+                    if (!this.tagsToMetricPointIndexDictionary.TryGetValue(sortedTags, out aggregatorIndex))
                     {
-                        aggregatorIndex = Interlocked.Increment(ref this.metricPointIndex);
+                        aggregatorIndex = ++this.metricPointIndex;
```

Comment:
I don't think ++/-- are atomic operations and we need the `Interlocked.Increment`. The reason is ++/-- compile to a read and then a set. Each are atomic individually, but they are not atomic together. Other threads can jump in between the read and set.

This...

```csharp
metricPointIndex++;
```

...compiles like...

```msil
    IL_0000: ldsfld       int32 class_name::metricPointIndex
    IL_0005: ldc.i4.1
    IL_0006: add
    IL_0007: stsfld       int32 class_name::metricPointIndex
```

Not 100% certain, but dangerous bet to lose 😄

Reply:
we are inside lock (for whole dictionary..), so ++ should be fine?

Reply:
Was chatting with @utpilla on slack, he said we dropped the `Interlocked.Increment` because we now have `lock (this.tagsToMetricPointIndexDictionary)`. I have different concerns with that. Before we had `lock (value2metrics)` which was locking on the child lookup. The reason for that is it reduced contention for updates with different keys. If we have this global lock we are going to add contention for updates across different keys. If the only reason for this was to remove the `Interlocked`, I think the old way was a better mechanism. Do our benchmarks test the same key/values over and over or is there a mix of uniqueness?

Reply:
The benchmarks always use the same set of keys, but they randomize the values for the respective keys.

Reply:
this PR removed the 2 level lookup, so we lose some niceness from the original approach. (wasted space to store keys over and over, more contention between new updates with different keys.)

Benchmarks *and* stress test should tell us if the benefits outweigh these concerns

Reply:
The main reason behind doing it was not to get rid of `Interlocked.Increment` but to get rid of sorting the key-value pairs on every update call. Separating the keys from values made it tough to identify the metricPoint index assigned to a tags combination when we tried to avoid sorting.